Use destroy over delete when deleting translation records

mpataki opened the pull request:
Ensures that rails callbacks get called on destroy.

timsly commented:
@mpataki are there any specific reasons why we need to use `destroy` over `delete`?

gr8bit commented:
For the sake of possible callbacks in Translation model I guess. Not the worst idea imho, even if it might impact performance as models are fetched, instantiated and deleted one by one...

mpataki commented:
Oof, sorry for the ulta-late reply on this. I'm just working out better github notification so things like this don't get lost in the noise.

Using `destroy` over `delete` opens up the option to use rails callbacks to react to the destroy event. In my case, using a fork of this repo which included this change, I used it to track these destroy events in one environment so I could verify and replicate them on others, but of course there could be many other uses.

As for the performance concern, I leave that to you guys. I know that in my use cases it never had any impact as we aren't really storing new translations on the fly, just from our internal admin panel as people add / update language, but I can't really speak for all users.

Thanks for having a look! I'll be more responsive in the future.

timsly commented:
@mpataki Thanks for explanation.
And I think this PR definitely make sense, but I would rather make it optional.
So to use `destroy` over `delete` you just need to configure it somewhere

mpataki commented:
That sounds reasonable. I'll put something together. Thanks!

timsly commented:
You can add some simple configuration code here: `lib/i18n/active_record.rb`
And in the end we can use it:
```ruby
I18n::ActiveRecord.configure do |config|
  config.cleanup_with_destroy = true
end
```

mpataki commented:
Ok, this should do it. I ended up adding the configuration to `I18n::Backend::ActiveRecord` instead of `I18n::ActiveRecord` as it seemed to be more in line with what you have going on here already, but I'm happy to re-work it so it's in `lib/i18n/active_record.rb` if you'd still prefer. I've also added some tests, version bumped, and updated the readme.

Let me know if there's anything further you'd like to see changed!

timsly commented on lib/i18n/backend/active_record.rb:
```diff
@@ -4,9 +4,18 @@
+      end
+
+      def self.config
+        @@config ||= Configuration.new
```
Let's use `@config` instead

mpataki replied:
This being a class variable, I don't think that's allowed by ruby. I get this:
```
NameError: `@config' is not allowed as a class variable name
```

timsly replied:
Let's go with `@_config` then

mpataki replied:
Yeah, same thing. Investigating a little further, it's [`class_variable_set`](http://apidock.com/ruby/Module/class_variable_set) that's throwing it, and otherwise it's allowed to be `@config` (one `@` instead of two). I was aiming to avoid exposing a writer for the config on `ActiveRecord`, but exposing one and using it over `class_variable_set` in the tests accomplishes what you're asking for, if you don't mind the setter being there.

I'll make that change and push so you can see it.

timsly commented:
@mpataki PR looks nice, i added one notice.
Can you please also squash all commits

timsly commented on lib/i18n/backend/active_record.rb:
```diff
@@ -4,9 +4,24 @@
+      autoload :Configuration, 'i18n/backend/active_record/configuration'
+
+      class << self
+        def configure(&block)
```
No need to pass `&block` here, since we using `yield` and `block_given?`

timsly commented on lib/i18n/backend/active_record.rb:
```diff
@@ -4,9 +4,24 @@
+          @config ||= Configuration.new
+        end
+
+        def config=(x)
```
What is the purpose for this setter?

mpataki replied:
Yeah I guess my comment from above in regards to your `@_config` suggestion got hidden. I'll paste it here:

Yeah, same thing. Investigating a little further, it's class_variable_set that's throwing it, and otherwise it's allowed to be @config (one @ instead of two). I was aiming to avoid exposing a writer for the config on ActiveRecord, but exposing one and using it over class_variable_set in the tests accomplishes what you're asking for, if you don't mind the setter being there.

I'll make that change and push so you can see it.

timsly replied:
@mpataki Ok, i got the problem and i think i know how to solve it.
So let's remove `config=` from here and to reset configuration in `teardown` we can use
```ruby
I18n::Backend::ActiveRecord.instance_variable_set :@config, I18n::Backend::ActiveRecord::Configuration.new
```

mpataki replied:
Woah, very cool - I hadn't considered `I18n::Backend::ActiveRecord` as itself being an instance of `Class`, and thus having its own instance variables. I learned something here, thanks man 👏!

I've made the change and squashed the commits again. Let me know if there's anything else.

timsly commented:
Cool, great work @mpataki, now we have configuration api. 
Thanks for contribution!
